hail/python/hailtop/utils/utils.py

```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
         thread_pool, lambda: fun(*args, **kwargs))
 
 
-class AsyncWorkerPool:
-    def __init__(self, parallelism):
-        self._sem = asyncio.Semaphore(parallelism)
-        self._count = 0
+async def gather(*coros, parallelism=10, return_exceptions=False):
+    gatherer = AsyncThrottledGather(*coros,
+                                    parallelism=parallelism,
+                                    return_exceptions=return_exceptions)
+    return await gatherer.wait()
+
+
+class AsyncThrottledGather:
+    def __init__(self, *coros, parallelism=10, return_exceptions=False):
+        self.count = len(coros)
+        self.n_finished = 0
+
+        self._queue = asyncio.Queue(maxsize=1000)
         self._done = asyncio.Event()
+        self._return_exceptions = return_exceptions
+        self._futures = []
+
+        self._results = [None] * len(coros)
+        self._error = None
```

Comment:
This should be a list and we should append to it. As written, one worker could set the error and trigger the done event, another worker could then overwrite the error with a new error and trigger the done event again, and then finally the wait thread wakes up and raises the second error. We can just always through the first exception, but I think it will be less confusion if we know it's always the first exception that gets raised.

Longer term, we might need to write a new exception that contains all the other exceptions. I'm not exactly sure how to do that, so let's punt on that.